src/box/lua/tuple.lua

```diff
@@ -356,18 +383,51 @@ ffi.metatype(tuple_iterator_t, {
     __tostring = function(self) return "<tuple iterator>" end;
 })
 
+local tuple_format_t = ffi.typeof('box_tuple_format_t')
+local tuple_format_serialize = internal.tuple.format.__serialize
+ffi.metatype(tuple_format_t, {
+    __gc = internal.tuple.format.__gc,
+    __ipairs = internal.tuple.format.__ipairs,
+    __pairs = internal.tuple.format.__pairs,
+    __index = function(_, key)
```

Comment:
Turns out we need cdata to implement the `__pairs` method because userdata objects don't support it in Lua 5.1

@Totktonada AFAIK cdata objects and FFI are disapproved by LuaJIT developers (IIRC Igor told me that). Moreover, cdata+FFI implementation looks more complex in this case. Do we really want to use cdata instead of userdata for tuple format objects because of `__pairs`? Maybe, we'd better add a plain method `pairs()` and implement `__pairs()` only when we update to Lua 5.2?

Reply:
Sorry, actually, I just reviewed the design document, at turned out that nobody insisted on `__pairs` — it was just my phantom memory about the discussion around `box.session.pairs`, where @unera (not sure if @Totktonada was present there) expressed a desire for this.

Reply:
I can't tell anything about formal rules here. I can bring some context and my opinion and let you decide.

## Tarantool

The `userdata` type looks as initially designed for defining objects with behavior written in C. The `cdata` type is more about interacting with existing C functions.

However, Tarantool has several functions that fill the gap: `luaL_cdef()`, `luaL_pushcdata()`, `luaL_tocpointer()`, `luaL_iscdata()`, `luaL_checkcdata()`, `luaL_ctypeid()`, `luaL_metatype()`, `luaL_setcdatagc()`.

There is no consistency in using `cdata` and `userdata` for our own objects.

## LuaJIT

There was a [proposal](https://github.com/LuaJIT/LuaJIT/pull/435) to add several functions to work with `cdata` from C to LuaJIT. It was declined by the LuaJIT author with the following reason:

> <...> mixing the classic Lua/C API and the FFI is never a good idea. The suggested new functions would give the wrong impression that this is safe or even reasonable to do.

## __pairs

Well, sure, there is the opposite side gap: `userdata` in Lua 5.1 can't affect `pairs()` using the `__pairs` metamethod.

## Assigning a metatable

There were a problem with lack of `ffi.metatype()` on the C side. So, some old code has a Lua part just to call this function (key_def, merger). It seems, it is solved by 4ca3953728c85c0ba77e411f0e7e8a2e183f1c35.

## Known opinions

Mike's Pall opinion is quotes above.

@igormunkin considers using of `cdata` for defining our own objects as improper usage of the type. He would like to better split LuaJIT and Tarantool source code by eliminating LuaJIT's internal APIs usages: while all the `luaL_pushcdata()` and friends are mostly built on top the internal API.

I don't know, whether Igor disapprove using `cdata` for new tarantool's types or not. I'll CC him.

## My opinion

I have no preference here, to be honest. I see no much difference from the readability point of view if the code is well shaped and the type specific new/is/check/push/gc functions incapsulate all the corresponding basic primitives.

## Conclusion

It seems, you should elaborate an agreement. I can't be a referee here.

// CC @igormunkin

Reply:
I prefer going the path of least confrontation: changed `box.tuple.format` objects to userdata.